File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const map = {
        ')': '(',
        '}': '{',
        ']': '['
    }
    const stack = [];
    
    for (const item of s) {
        if (item in map) {
            if (stack.length > 0 && stack[stack.length - 1] === map[item]) {
                stack.pop();
            } else {
                return false;
            }
        } else {
            stack.push(item);
        }
    }
    return stack.length === 0;
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        bracketDict = {
            ')': '(',
            ']': '[',
            '}': '{'
        }
        stack = []
        
        for bracket in s:
            if bracket in bracketDict:
                if len(stack) > 0 and stack[-1] == bracketDict[bracket]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(bracket)
                
        return True if not stack else False